### DISCUSSION TITLE
✨ Add environment variable support to arguments

### DISCUSSION THREAD

**Discussion by rawkode** [dagger/dagger#12170] [Ideas]
### What are you trying to do?

It would be fantastic if we could allow argument annotations on functions, this would allow pulling the environment from .envrc or .env files 

### Why is this important to you?

Each service in my mono repo configures itself with direnv. Currently I'm duplicating a lot to make workflows work locally and in Dagger. 

### Related

- https://github.com/dagger/dagger/issues/9584

**Comment by dragoonis** [comment]
I'm curious @rawkode, is this what you're referring to?

I had to write this in PHP, as a workaround. It would be great not to have to do this

``` php
    private function loadEnvArgs(Container $container, Directory $source): Container
    {
        $envContents = $source->file('.env')->contents();
        $envVars = parse_ini_string($envContents);

        foreach($envVars as $envVarKey => $envVarValue) {
            $container = $container->withEnvVariable($envVarKey, $envVarValue);
        }

        return $container;
    }

**Comment by rawkode** [comment]
No. My request is for the ability to skip using `--value env:SOME` when invoking Dagger CLI.

**Comment by Vad1mo** [comment]
To keep things excplicity, I would explicitly speicify environment variables that need to be imported to dagger, And I would have to types `importEnvVar(NAME)` and `requireEnvVar(NAME)`. In the second case Dagger will fail if the env var does not exist or is empty.

**Comment by jasonmccallister** [comment]
@rawkode to confirm your use case, you are referring to adding a feature the allow module/function arguments that are currently passed as flags to support environment variables; similar to tools like [urfav/cli](https://cli.urfave.org/v3/examples/flags/#values-from-the-environment) environment variables?

If I was using the [example from the docs](https://docs.dagger.io/api/arguments#default-values) it would look something like this?

```go
package main

import (
	"context"
	"fmt"
)

type MyModule struct{}

func (m *MyModule) Hello(
	ctx context.Context,
	// +optional
	// +default="world"
        // +environmentVariable=MY_NAME
	name string,
) (string, error) {
	return fmt.Sprintf("Hello, %s", name), nil
}
```

That way you can simplify the dagger calls:

```bash
$ export MY_NAME=Jason
 
$ dagger call hello

$ Hello, Jason
```

**Comment by rawkode** [comment]
Yep!

**Comment by jasonmccallister** [comment]
Related https://github.com/dagger/dagger/issues/9584

**Comment by bupd** [comment]
I have written a shell script to workaround this in the meantime.

```sh
#!/bin/sh

envFile="/envFile"

# Check if the file exists
if [ ! -f "$envFile" ]; then
  echo "Environment file not found: $envFile"
  exit 1
fi

# Read the file and export each line as an environment variable
while IFS='=' read -r key value; do
  # Ignore empty lines or comments
  if [ -z "$key" ] || [ "${key:0:1}" = "#" ]; then
    continue
  fi

  # Export the environment variable
  export "$key"="$value"

  # print the variable to verify it's set
  echo "Set $key=$value"
done < "$envFile"

# Execute the command (passed as $1 arg)
echo "Executing: $1"
eval "$1"
```
